### PR TITLE
[#48044] Use Location picker modal to insert project folder

### PR DESCRIPTION
**Pull request by Kharonus**
- https://community.openproject.org/work_packages/48044
- amend file/location picker modal to contain error states
- add location picker functionality to project storages form

### CODE REVIEW

File: docs/api/apiv3/components/schemas/project_storage_model.yml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+properties:
+  _type:
+    type: string
+    enum:
+      - ProjectStorage
+  id:
+    type: integer
+    description: The project storage's id
+    minimum: 1
+  projectFolderMode:
```
Comment (apfohl) [inline docs/api/apiv3/components/schemas/project_storage_model.yml:19]: I think `ProjectFolderId` is missing.

File: docs/api/apiv3/paths/project_storages.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      example: '[{ "project_id": { "operator": "=", "values": ["42"] }}, { "storage_id": { "operator": "=", "values": ["1337"] }}]'
+  responses:
+    '200':
+      description: OK
+      content:
+        application/hal+json:
+          schema:
+            $ref: '../components/schemas/project_storage_collection_model.yml'
+    '400':
+      description: Returned if the any given filter is invalid.
```
Comment (apfohl) [inline docs/api/apiv3/paths/project_storages.yml]: minus `the` or `any`

File: modules/storages/app/controllers/storages/admin/projects_storages_controller.rb [hunk 3]
```diff
@@ -142,6 +145,50 @@ def destroy
...
+          set_permissions_command = Storages::Peripherals::StorageRequests.new(storage:).set_permissions_command
+          set_permissions_command.result.call(folder:, permissions:).match(
+            on_success: ->(_) { flash[:notice] = 'Permissions were successfuly updated on the NextCloud side' }, # rubocop:disable Rails/I18nLocaleTexts
+            on_failure: ->(error) { flash[:error] = "Error: #{error}" }
+          )
+        end,
+        on_failure: ->(error) do
+          flash[:error] = "Error: #{error}"
+        end
+      )
```
Comment (akabiru) [inline modules/storages/app/controllers/storages/admin/projects_storages_controller.rb]: Worth a merge with dev, seems this was removed in https://github.com/opf/openproject/pull/12697

File: frontend/src/app/shared/components/storages/file-picker-base-modal.component.ts/file-picker-base-modal.component.ts [hunk 4]
```diff
@@ -138,25 +142,29 @@ export abstract class FilePickerBaseModalComponent extends OpModalComponent impl
 
   protected changeLevel(ancestor:IStorageFile):void {
     this.cancelCurrentLoading();
-    this.loading$.next(true);
+    this.loading$.next('loading');
 
     this.loadingSubscription = this.storageFilesResourceService
       .files(makeFilesCollectionLink(this.storage._links.self, ancestor.location))
+      .pipe(catchError((error) => {
```
Comment (apfohl) [inline frontend/src/app/shared/components/storages/file-picker-base-modal.component.ts/file-picker-base-modal.component.ts]: You could catch the error in the subscribe, rather then in a side effect.

**Review by akabiru — APPROVED** [review]
Looks great for my POV, and manual tests pass as well! 👍🏾

File: frontend/src/app/shared/components/storages/file-picker-modal/file-picker-modal.component.html [hunk 5]
```diff
@@ -40,16 +44,26 @@
...
+        <span class="spot-body-small op-file-picker--info-sub-text" [textContent]="text.content.emptyHint"></span>
+      </div>
+    </ng-template>
+
+    <ng-template #notFoundTemplate>
+      <div class="spot-body-big spot-body-big_bold op-file-picker--info">
+        <span class="spot-icon spot-icon_remove-link op-file-picker--info-icon"></span>
+        <span [textContent]="text.content.noConnection(storageType)"></span>
+        <span class="spot-body-small op-file-picker--info-sub-text"
+              [textContent]="text.content.noConnectionHint(storageType)"></span>
```
Comment (apfohl) [inline frontend/src/app/shared/components/storages/file-picker-modal/file-picker-modal.component.html:65]: `FilePickerModal` and `LocationPickerModal` have somehow similar templates implemented. Could this be factored out?

**Review by apfohl — CHANGES_REQUESTED** [review]

File: modules/storages/app/controllers/storages/admin/projects_storages_controller.rb (hunk 3 above)
Comment (Kharonus) [reply modules/storages/app/controllers/storages/admin/projects_storages_controller.rb]: I removed it, but I really have no idea, why this code was still in.

File: docs/api/apiv3/components/schemas/project_storage_model.yml (hunk 1 above)
Comment (Kharonus) [reply docs/api/apiv3/components/schemas/project_storage_model.yml:19]: `projectFolderId` was removed and replaced with the link `projectFolder`.

File: docs/api/apiv3/paths/project_storages.yml (hunk 2 above)
Comment (Kharonus) [reply docs/api/apiv3/paths/project_storages.yml]: removed

File: frontend/src/app/shared/components/storages/file-picker-base-modal.component.ts/file-picker-base-modal.component.ts (hunk 4 above)
Comment (Kharonus) [reply frontend/src/app/shared/components/storages/file-picker-base-modal.component.ts/file-picker-base-modal.component.ts]: moved and embraced new rxjs subscribe ^^

File: frontend/src/app/shared/components/storages/file-picker-modal/file-picker-modal.component.html (hunk 5 above)
Comment (Kharonus) [reply frontend/src/app/shared/components/storages/file-picker-modal/file-picker-modal.component.html:65]: they have, but until now it was easier for me to just use two templates. They do share a lot, but there are some differences in the action bar and with the list items. template sharing in angular also kind of sucks, until someone can show me a nice way how to do it :D

**Review by apfohl — APPROVED** [review]

**Pull request merged**